### DISCUSSION TITLE
Understand Node LTS versioning

### DISCUSSION THREAD

**Discussion by mwgamble** [renovatebot/renovate#18561] [Request Help] [answered]
### What would you like Renovate to be able to do?

I would like to be able to tell Node.js to never update to a non-LTS release of Node.js. Currently to achieve this I need a package rule for `node` that sets `allowedVersions` to `<19`, and I need to remember to update this everywhere once a year. This is toil that could be eliminated by proper support from Renovate.

While it doesn't currently have a concept of "LTS", Nginx is another piece of software that utilises a similar versioning scheme. Odd-numbered releases (1.21, 1.23, etc) are "mainline" releases not intended for production, while even-numbered releases (1.20, 1.22, etc) are intended for production. I don't care anywhere near as much about this (most users aren't deploying nginx themselves where I work), but it may be useful context when designing a solution.

### If you have any ideas on how this should be implemented, please tell us here.

The easiest solution would probably be to introduce a new versioning strategy for Node.js specifically that understands the difference between LTS and non-LTS releases, and to automatically apply this versioning strategy to anything that looks like Node.js (like the official images on Docker Hub, for example).

### Is this a feature you are interested in implementing yourself?

No

**Comment by rarkins** [comment, marked as answer]
Node versioning already exists in Renovate

**Reply by rgl** [reply]
@rarkins, can you please share an example on how to apply it?

For example, at https://github.com/rgl/my-ubuntu-ansible-playbooks/pull/22/files, renovate is trying to upgrade to non-lts version, I'm using:

```yaml
# renovate: datasource=node depName=node
nodejs_version: '16.18.0'
```

Where it will be handled by [a custom regexManagers](https://github.com/rgl/my-ubuntu-ansible-playbooks/blob/main/renovate.json5) which extracts the properties:

```
datasource=node
depName=node
currentValue=16.18.0
versioningTemplate=semver
```

Is there a way configure the version to be `lts` (which is something that [seems to be supported by the node datasource in the form of the `isStable` property](https://github.com/renovatebot/renovate/blob/e4dbd4ad491f656aaa7f9da9d4dae2c18622a1e6/lib/modules/datasource/node/index.ts#L53))?

I'm not seeing how this would work just by setting `versioningTemplate=node`, as I'm not defined it to use `lts` anywhere.

**Reply by rarkins** [reply]
Try setting versioning=node too, in your log it shows semver

Otherwise extract your example out into a reproduction repo that's easy to debug

**Reply by viceice** [reply]
you shouldn't assing a versioning explicit, the node datasource will use node versioning automatically if no userdefined is found

https://github.com/renovatebot/renovate/blob/4fc1019cb6775bd4394e200ff57e65f541b6f91f/lib/modules/datasource/node/index.ts#L3
https://github.com/renovatebot/renovate/blob/4fc1019cb6775bd4394e200ff57e65f541b6f91f/lib/modules/datasource/node/index.ts#L20

**Reply by viceice** [reply]
node versioning only allows lts updates by default

https://github.com/renovatebot/renovate/blob/4fc1019cb6775bd4394e200ff57e65f541b6f91f/lib/modules/versioning/node/index.ts#L51-L60

**Reply by rarkins** [reply]
Good point, removing `"versioningTemplate": "semver"` might be enough to get it to default to `node` versioning

**Reply by rgl** [reply]
I've changed the `versioningTemplate` to `node`, but it still tries to upgrade to v18.

Does this look ok?

```json
{
  "packageFile": "roles/nodejs/vars/main.yml",
  "dep": {
    "depName": "node",
    "currentValue": "16.18.0",
    "datasource": "node",
    "versioning": "node",
    "extractVersion": "^v?(?<version>.+)$",
    "replaceString": "# renovate: datasource=node depName=node versioning=node\nnodejs_version: '16.18.0'\n",
    "depIndex": 0,
    "updates": [
      {
        "bucket": "major",
        "newVersion": "18.12.0",
        "newValue": "18.12.0",
        "releaseTimestamp": "2022-10-25T00:00:00.000Z",
        "newMajor": 18,
        "newMinor": 12,
        "updateType": "major",
        "branchName": "renovate/node-18.x"
      }
    ],
    "warnings": [],
    "sourceUrl": "https://github.com/nodejs/node",
    "homepage": "https://nodejs.org",
    "currentVersion": "16.18.0",
    "isSingleVersion": true,
    "fixedVersion": "16.18.0"
  }
}
```

**Reply by rarkins** [reply]
v18 is LTS

**Reply by rgl** [reply]
Oh, I'm so sorry! I've completely missed the 2022-10-25 release date!

Everything is working as expected. Thank you!
